plans/example-browser-node/browser/index.js

```diff
@@ -0,0 +1,105 @@
+const { chromium, firefox, webkit } = require('playwright')
+const { exit } = require('process')
+const { spawn } = require('child_process')
+
+const spawnServer = require('../server')
+
+;(async () => {
+  spawnServer(8080)
+
+  let browser
+  try {
+    const browserDebugPort = process.env.BROWSER_DEBUG_PORT || 9222
+
+    switch (process.env.BROWSER_KIND || 'chromium') {
+      case 'chromium':
+        console.log(`launching chromium browser with exposed debug port: ${browserDebugPort}`)
+        browser = await chromium.launch({
+          args: [
+            '--remote-debugging-address=0.0.0.0',
+            `--remote-debugging-port=${browserDebugPort}`
+          ]
+        })
+        break
+
+      case 'firefox':
+        const localBrowserDebugPort = Number(browserDebugPort) + 1
+        console.log(`launching firefox browser with exposed debug port: ${browserDebugPort} (local ${localBrowserDebugPort})`)
+        browser = await firefox.launch({
+          args: [
+            `-start-debugger-server=${localBrowserDebugPort}`
+          ]
+        })
+
+        console.log('launching tcp proxy to expose firefox debugger for remote access')
+        const tcpProxy = spawn(
+          'socat', [
+            `tcp-listen:${browserDebugPort},bind=0.0.0.0,fork`,
+            `tcp:localhost:${localBrowserDebugPort}`
+          ]
+        )
+        tcpProxy.stdout.on('data', (data) => {
+          console.log(`firefox debugger: tcpProxy: stdout: ${data}`)
+        })
+
+        tcpProxy.stderr.on('data', (data) => {
+          console.error(`firefox debugger: tcpProxy: stderr: ${data}`)
+        })
+
+        break
+
+      case 'webkit':
+        console.log('launching webkit browser (remote debugging not yet supported)')
+        browser = await webkit.launch()
+        break
+    }
+
+    const page = await browser.newPage()
+
+    page.on('console', (message) => {
+      const loc = message.location()
+      console.log(`[${message.type()}] ${loc.url}@L${loc.lineNumber}:C${loc.columnNumber}: ${message.text()} — ${message.args()}`)
+    })
+
+    console.log('prepare page window (global) environment')
+    await page.addInitScript((nodeEnv) => {
+      const env = {}
+      for (const k in nodeEnv) {
+        if (k.startsWith('TEST_')) {
```

Comment:
This might work now, but it will bite us eventually,

Since the testground SDK is the one official owner of reading and parsing testground envs,
could you make sure we rely on it to load & forward the env to the browser runtime?

Adding a serialize / deserialize feature in https://github.com/testground/sdk-js/blob/master/src/runtime/params.js might be enough.

Reply:
What exactly might be us eventually? Also what would the serialize/deserialize feature solve here? I am open to anything, I'm asking as I am not following exactly what you're referring to.

Reply:
Could you please clarify on this? A meeting isn't needed, would be useful if you can simply respond.

Reply:
@GlenDC this code relies on the fact that every variable is prefixed with `TEST_`, which is not enforced anywhere, it's an implicit specification.

The reliable way to access Testground's environment variables is to use the SDK. The corresponding namespace is defined in: https://github.com/testground/sdk-js/blob/master/src/runtime/params.js.

I want to make sure we rely on the SDK to retrieve these values and don't rely on implicit constraints like "every env variable is prefixed with `TEST_`, which might become untrue anytime).

Reply:
Added functionality in sdk-js PR to be able to do this: https://github.com/testground/sdk-js/commit/06df3ce6c9c904d52fa6d081aa1ffd204b5de69a

This I have implemented also in this PR already (the integrated functionality): https://github.com/testground/testground/pull/1502/commits/d7f0b02e72293586268dd127c2bf11c6bd4616ac